### PR TITLE
Add optional `required` parameter to VaultSecretValue

### PR DESCRIPTION
**Pull request by bfaulk96**
Add an optional `required` parameter to VaultSecretValues so that secrets can quietly be empty if `required: false` is set.

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira (N/A)
- [x] Link to relevant pull requests, esp. upstream and downstream changes (N/A)
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

Edit: Further explanation
while `failIfNotFound` configuration option exists, this is for handling when a _secret_ is not found. This PR is to more granularly allow optional values within a secret, for example:

```groovy
  def secrets = [
    [
      path: 'some/path/in/vault',
      secretValues: [
       [envVar: 'MANDATORY_THING', vaultKey: 'mandatoryThing', isRequired: true]
       [envVar: 'OPTIONAL_THING', vaultKey: 'optionalThing', isRequired: false],
      ]
    ]
  ]
```

### CODE REVIEW

**Comment by bfaulk96** [comment]
Admittedly, I haven't worked with Java since college, and I don't really know the structure of this project to add unit tests right now. I can look into adding tests for this later, but if they are simple and someone knows what would be needed, that would be awesome as well.

**Comment by Vladyslav-Miletskyi** [comment]
@bfaulk96 damn, I wanted to do add the same functionality :) great job 👍🏻 


@jetersen is there a chance to have this merged any time soon

File: src/main/java/com/datapipe/jenkins/vault/model/VaultSecretValue.java
Comment (jetersen) [inline src/main/java/com/datapipe/jenkins/vault/model/VaultSecretValue.java:55]: Why is this here? Not sure this should be the default
```suggestion
```

**Review by jetersen — CHANGES_REQUESTED** [review]
Your missing the UI part of it.

https://github.com/jenkinsci/hashicorp-vault-plugin/blob/master/src/main/resources/com/datapipe/jenkins/vault/model/VaultSecretValue/config.jelly

Otherwise looks fine, though not sure why it is necessary when you have getFailIfNotFound

https://github.com/jenkinsci/hashicorp-vault-plugin/blob/a35e28a212c2ca0f51725c643b7e338a4f90cbdf/src/main/java/com/datapipe/jenkins/vault/VaultAccessor.java#L233-L235

Actually this would throw before isRequired is checked for false so kind of a moot point.

File: src/main/java/com/datapipe/jenkins/vault/model/VaultSecretValue.java
Comment (bfaulk96) [reply src/main/java/com/datapipe/jenkins/vault/model/VaultSecretValue.java:55]: The current default behavior is that all vault secrets are required (which is why I created this PR)... Figured I'd add this default as to not change current behavior, only add an additional option. I personally don't care either way though.

**Comment by bfaulk96** [comment]
> Your missing the UI part of it.
> 
> https://github.com/jenkinsci/hashicorp-vault-plugin/blob/master/src/main/resources/com/datapipe/jenkins/vault/model/VaultSecretValue/config.jelly
> 
> Otherwise looks fine, though not sure why it is necessary when you have getFailIfNotFound
> 
> https://github.com/jenkinsci/hashicorp-vault-plugin/blob/a35e28a212c2ca0f51725c643b7e338a4f90cbdf/src/main/java/com/datapipe/jenkins/vault/VaultAccessor.java#L233-L235
> 
> Actually this would throw before isRequired is checked for false so kind of a moot point.

`failIfNotFound` is for finding the secret itself, not for getting values in the secret.

i.e.:
```groovy
  def secrets = [
    [
      path: 'some/path/in/vault', // <--`failIfNotFound` option
      secretValues: [
       [envVar: 'OPTIONAL_THING', vaultKey: 'optionalThing'], // Currently these throw an error if not in Vault, which is what this PR aims to fix
      ]
    ]
  ]
```

File: src/main/java/com/datapipe/jenkins/vault/model/VaultSecretValue.java
Comment (bfaulk96) [reply src/main/java/com/datapipe/jenkins/vault/model/VaultSecretValue.java:55]: This is in reference to the values, not the secrets themselves.
`FailIfNotFound` fails if the secrets are not found
`isRequired` on the secret values is specifically when a secret does exist, but an optional value does not.

**Comment by bfaulk96** [comment]
> Your missing the UI part of it.
> 
> https://github.com/jenkinsci/hashicorp-vault-plugin/blob/master/src/main/resources/com/datapipe/jenkins/vault/model/VaultSecretValue/config.jelly
> 
> Otherwise looks fine, though not sure why it is necessary when you have getFailIfNotFound
> 
> https://github.com/jenkinsci/hashicorp-vault-plugin/blob/a35e28a212c2ca0f51725c643b7e338a4f90cbdf/src/main/java/com/datapipe/jenkins/vault/VaultAccessor.java#L233-L235
> 
> Actually this would throw before isRequired is checked for false so kind of a moot point.

@jetersen as requested, the UI portions have been added

File: src/main/java/com/datapipe/jenkins/vault/model/VaultSecretValue.java
Comment (jetersen) [inline src/main/java/com/datapipe/jenkins/vault/model/VaultSecretValue.java:62]: You can only have one `@DataBoundConstructor`
```suggestion
```

You need to add a `@DataBoundSetter` for `isRequired

File: src/main/resources/com/datapipe/jenkins/vault/model/VaultSecretValue/config.jelly
Comment (jetersen) [inline src/main/resources/com/datapipe/jenkins/vault/model/VaultSecretValue/config.jelly:14]: [JENKINS-55787](https://issues.jenkins.io/browse/JENKINS-55787)

```suggestion
  <f:entry field="isRequired">
    <f:checkbox title="Required?" />
  </f:entry>
```

**Review by jetersen — CHANGES_REQUESTED** [review]

File: src/main/java/com/datapipe/jenkins/vault/model/VaultSecretValue.java
Comment (bfaulk96) [reply src/main/java/com/datapipe/jenkins/vault/model/VaultSecretValue.java]: Sorry, shooting in the dark on most of this, but doing my best with it 😆

**Comment by bfaulk96** [comment]
@jetersen Any update on this?

**Comment by bfaulk96** [comment]
bump

File: src/main/java/com/datapipe/jenkins/vault/model/VaultSecretValue.java
Comment (jetersen) [inline src/main/java/com/datapipe/jenkins/vault/model/VaultSecretValue.java:56]: After testing this should preserve the old behavior
```suggestion
        this.isRequired = true;
```

**Review by jetersen — APPROVED** [review]

File: src/main/java/com/datapipe/jenkins/vault/model/VaultSecretValue.java
Comment (jetersen) [reply src/main/java/com/datapipe/jenkins/vault/model/VaultSecretValue.java:56]: Now that I think about this will most likely always force it to true.

We should follow the pattern for defaults:
https://github.com/jenkinsci/notify-events-plugin/blob/3fbfb7b65d9dd72140692fc51d3da2911941a2a9/src/main/java/events/notify/NotifyEventsBuilder.java#L144-L149

Comment (bfaulk96) [reply src/main/java/com/datapipe/jenkins/vault/model/VaultSecretValue.java]: @jetersen done!

File: src/main/java/com/datapipe/jenkins/vault/model/VaultSecretValue.java
Comment (jetersen) [inline src/main/java/com/datapipe/jenkins/vault/model/VaultSecretValue.java:42]: Should be a public final static on the DescriptorImpl.

File: src/main/java/com/datapipe/jenkins/vault/model/VaultSecretValue.java
Comment (jetersen) [inline src/main/java/com/datapipe/jenkins/vault/model/VaultSecretValue.java:58]: This would still force to always default to true.

The correct behavior is to set the default on the .jelly file :)
https://github.com/jenkinsci/compuware-topaz-for-total-test-plugin/blob/e3c8508809e1befad31e6908785322a06fbbe944/src/main/resources/com/compuware/jenkins/totaltest/TotalTestCTBuilder/config.jelly#L65

**Review by jetersen — COMMENTED** [review]
Sorry for not being 100% clear on the standards :)

File: src/main/java/com/datapipe/jenkins/vault/model/VaultSecretValue.java
Comment (bfaulk96) [reply src/main/java/com/datapipe/jenkins/vault/model/VaultSecretValue.java:58]: A couple quick questions on this... Does this mean I shouldn't have the lines
```java
this.isRequired = DescriptorImpl.DEFAULT_IS_REQUIRED;
```
at all? Or I should just have those **and** the jelly file changes?

Additionally, just for my own curiosity, how does all of this work when calling the vault plugin from a declarative pipeline as opposed to using the global GUI config (which I assume is what the .jelly files are for)?

Comment (jetersen) [reply src/main/java/com/datapipe/jenkins/vault/model/VaultSecretValue.java]: This should not be there at all in the constructor.

The default should be added to the private field definition: see https://github.com/jenkinsci/compuware-topaz-for-total-test-plugin/blob/e3c8508809e1befad31e6908785322a06fbbe944/src/main/java/com/compuware/jenkins/totaltest/TotalTestCTBuilder.java#L136

**Review by jetersen — APPROVED** [review]
LGTM

Sorry for all the back and forth 😅

**Comment by bfaulk96** [comment]
> LGTM
> 
> Sorry for all the back and forth 😅

Ha, no worries. I kind of figured it might happen as I was trying to make changes in a repo I didn't understand well 😅 
Thanks for the help!

Do you know why junit is failing in the build steps?
It wasn't doing that a while ago, so I'm a little confused by it

**Comment by bfaulk96** [comment]
Also, I'm not familiar with your team/org's merge process – but I don't have write access to this repo so I cannot merge it

**Comment by jetersen** [comment]
Merging it up to the maintainers of the repo :) Anyone can become a maintainer.

https://www.jenkins.io/doc/developer/plugin-governance/adopt-a-plugin/

**Comment by jetersen** [comment]
Jenkins CI seems to be sick :) However GitHub Action passes so I think it is fine.

**Comment by bfaulk96** [comment]
> Merging it up to the maintainers of the repo :) Anyone can become a maintainer.
> 
> https://www.jenkins.io/doc/developer/plugin-governance/adopt-a-plugin/

Welp, upon trying to request access I get notifications saying I can't send emails to either `jenkinsci-users` or `jenkinsci-dev` mailing list.

**Comment by jetersen** [comment]
I wasn't suggesting you become a maintainer of the plugin, unless your willing 😄

**Comment by bfaulk96** [comment]
Ah, alright. Due to my limited knowledge of Java, I guess I'll leave merging this to the _current_ maintainers 😅

**Comment by b1zzu** [comment]
Hi, I would really like to use this feature, when do you think it will be released?

**Comment by jetersen** [comment]
Let me setup continuos delivery to speed up the release process!

**Comment by jetersen** [comment]
@b1zzu released as https://github.com/jenkinsci/hashicorp-vault-plugin/releases/tag/336.v182c0fbaaeb7

**Pull request merged**
